### PR TITLE
First prototype

### PR DESCRIPTION
**Pull request by icoluccio**
## Summary
- Added the base implementation for wor-paginate that consists of:
  - The main paginate module that finds the correct adapter to use and responds to the user in a standard way. All this is done by requiring the module and calling render_paginate instead of render 
  - Adapters: each adapter knows how to handle a paginable object so that the formatter can output the standardized json
  - Formatter: implements conversion into json in a uniform way for anything that can be paginated
- Fixed lots of gem structure things

### CODE REVIEW

File: lib/wor/paginate/adapters/will_paginate_already_paginated.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+module Wor
+  module Paginate
+    module Adapters
+      class WillPaginateAlreadyPaginated
+        def adapt?(content)
+          # Methods will_paginate adds to ActiveRecord relations:
+          ### [:current_page, :total_entries, :total_entries=, :find_last, :current_page=,
+          ### :scoped, :total_pages, :next_page, :previous_page, :out_of_bounds?]
+          %i[previous_page out_of_bounds? total_entries=
```
Comment (hdf1986) [inline lib/wor/paginate/adapters/will_paginate_already_paginated.rb]: Sadly this seems to be the only way to detect the pagination gems :'(

File: lib/wor/paginate/paginate.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+                       Adapters::Iterable.new]
+    end
+
+    def render_paginated(content)
+      render json: paginate(content)
+    end
+
+    def paginate(content)
+      adapter = find_adapter_for_content(content)
+      throw "No pagination adapter for class #{content.class}" unless adapter.present?
```
Comment (hdf1986) [inline lib/wor/paginate/paginate.rb]: Can we raise a custom exception here?

File: spec/dummy/app/controllers/authentication_controller.rb [hunk 3]
```diff
@@ -0,0 +1,4 @@
+class AuthenticationController < ApplicationController
+  include Wor::Authentication::SessionsController
```
Comment (hdf1986) [inline spec/dummy/app/controllers/authentication_controller.rb]: We need this? :P

File: spec/dummy/app/controllers/dummy_models_controller.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+class DummyModelsController < ApplicationController
+  protect_from_forgery with: :exception
```
Comment (hdf1986) [inline spec/dummy/app/controllers/dummy_models_controller.rb]: We already have this in the application controller

File: spec/dummy_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+      it 'responds with total_count' do
+        response_json = JSON.parse(response.body)
+        expect(response_json['total']).to be 7
+      end
+
+      it 'responds with page' do
+        response_json = JSON.parse(response.body)
+        expect(response_json['page']).to be 0
+      end
```
Comment (hdf1986) [inline spec/dummy_controller_spec.rb]: Can we add tests for the content ?

File: wor-paginate.gemspec [hunk 6]
```diff
@@ -1,26 +1,29 @@
...
+$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
 require "wor/paginate/version"
 require 'date'
 
 Gem::Specification.new do |s|
   s.name        = "wor-paginate"
   s.version     = Wor::Paginate::VERSION
   s.platform    = Gem::Platform::RUBY
-  s.date        = Date.today
+  s.date        = '2017-06-23'
```
Comment (hdf1986) [inline wor-paginate.gemspec]: Hmm Date.today sounds better, this way when we build the gem the date updates by itself

File: wor-paginate.gemspec [hunk 7]
```diff
@@ -1,26 +1,29 @@
...
   s.license     = "MIT"
 
-  s.files = Dir["{app,config,db,lib}/**/*", "MIT-LICENSE", "Rakefile", "README.md"]
+  s.files = `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec)/}) }
   s.require_paths = ['lib']
 
 
   s.add_dependency 'railties', '>= 4.1.0', '< 5.1'
+  s.add_dependency 'rails', '>= 4.0'
+  s.add_dependency 'kaminari-activerecord'
```
Comment (hdf1986) [inline wor-paginate.gemspec]: We are going to finally add the dependency?

File: wor-paginate.gemspec [hunk 8]
```diff
@@ -31,4 +34,6 @@ Gem::Specification.new do |s|
   s.add_development_dependency 'codeclimate-test-reporter', '~> 1.0.0'
   s.add_development_dependency 'generator_spec'
   s.add_development_dependency 'simplecov'
+  s.add_development_dependency 'timecop'
```
Comment (hdf1986) [inline wor-paginate.gemspec]: I think that adding rails is enough for this

File: wor-paginate.gemspec (hunk 7 above)
Comment (icoluccio) [inline wor-paginate.gemspec]: Nope :D

File: wor-paginate.gemspec (hunk 8 above)
Comment (icoluccio) [inline wor-paginate.gemspec]: Plus we don't even need timecop so far

File: .gitignore [hunk 9]
```diff
@@ -1,7 +1,331 @@
...
+# Windows Installer files
+*.cab
+*.msi
+*.msm
+*.msp
+
+# Windows shortcuts
+*.lnk
+
+# End of https://www.gitignore.io/api/ruby,rubymine,rails,emacs,vim,sublimetext,osx,macos,linux,windows
```
Comment (icoluccio) [inline .gitignore]: Just in case someone else reads this: I got it from another wor gem. We can definitely make this simpler though

**Pull request merged**